hydra-node/src/Hydra/HeadLogic.hs

```diff
@@ -560,24 +644,237 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+    case mExistingDecommitTx of
+      Nothing ->
+        case applyTransactions currentSlot localUTxO [decommitTx] of
+          Right utxo' -> cont utxo'
+          Left (_, err)
+            | ttl > 0 ->
+                wait $ WaitOnNotApplicableDecommitTx decommitTx
+            | otherwise ->
+                cause . ClientEffect $
+                  ServerOutput.DecommitInvalid
+                    { headId
+                    , decommitInvalidReason =
+                        ServerOutput.DecommitTxInvalid
+                          { confirmedUTxO
+                          , decommitTx
+                          , validationError = err
+                          }
+                    }
+      Just existingDecommitTx
+        | ttl > 0 ->
+            wait $ WaitOnNotApplicableDecommitTx decommitTx
+        | otherwise ->
+            -- REVIW: cause . ClientEffect $ ServerOutput.DecommitInvalid
+            Error $ RequireFailed $ DecommitTxInFlight{decommitTx = existingDecommitTx}
+
+  maybeEmitSnapshot =
+    if isLeader parameters party nextSn
+      then cause (NetworkEffect (ReqSn version nextSn (txId <$> localTxs) (Just decommitTx)))
+      else noop
+
+  Environment{party} = env
+
+  Ledger{applyTransactions} = ledger
+
+  Snapshot{number} = getSnapshot confirmedSnapshot
+
+  nextSn = number + 1
+
+  CoordinatedHeadState{decommitTx = mExistingDecommitTx, confirmedSnapshot, localTxs, localUTxO, version} = coordinatedHeadState
+
+  confirmedUTxO = case confirmedSnapshot of
+    InitialSnapshot{initialUTxO} -> initialUTxO
+    ConfirmedSnapshot{snapshot = Snapshot{utxo}} -> utxo
+
+  OpenState
+    { headId
+    , parameters
+    , coordinatedHeadState
+    , currentSlot
+    } = openState
+
+-- ** Decrementing funds from the Head
+
+-- | Observe a decommit transaction. If the outputs of the pending decommit tx
+-- are equal to the latest confirmed UTxO to decommit, then we consider the
+-- decommit valid, and we take the funds out of the head and remove the
+-- decommit tx in flight.
+-- Finally, if the client observing happens to be the leader, then a new
+-- ReqSn is broadcasted.
+--
+-- __Transition__: 'OpenState' → 'OpenState'
+onOpenChainDecrementTx :: IsTx tx => Environment -> OpenState tx -> Outcome tx
+onOpenChainDecrementTx Environment{party} openState
+  | -- Spec: if outputs(txω) = 𝑈ω
+    -- REVIEW: should we get Uω from observation instead of local state?
+    (utxoFromTx <$> decommitTx) == utxoToDecommit =
+      -- Spec: txω ← ⊥
+      --       vˆ  ← v
+      newState DecommitFinalized
+        <> cause (ClientEffect $ ServerOutput.DecommitFinalized{headId})
+        -- Spec: if ŝ = S⁻.s ∧ leader(S⁻.s + 1) = i
+        --         multicast (reqSn, vˆ, S⁻.s + 1, T̂ , txω )
+        & maybeEmitSnapshot
+  | otherwise = noop
+ where
+  version' = version + 1
+
+  partyIsLeader = isLeader parameters party nextSn && not (null localTxs)
+
+  maybeEmitSnapshot outcome =
+    if seenSn == confirmedSn && partyIsLeader
```

Comment:
Fix also in spec: only snapshot if there is something to snapshot (i.e. localTxs is not null)

Reply:
@v0d1ch @ffakenz this is what you raised on slack. I realize that this is fine in the spec as-is: The snapshot leader would request a snapshot with the `decommitTx` being cleared (and any `localTxs` if there are any). Why? This makes the next decommit possible.

Reply:
I've done this now here: https://github.com/input-output-hk/hydra/commit/1f166b78e5e16c2be1758151127bfc2aac45204a